File: web/src/github.ts
const encoder = new TextEncoder();

export interface GitHubUser {
  login: string;
}

export interface Installation {
  id: number;
  account: {
    login: string;
    type: "Organization" | "User";
  };
}

export interface Repository {
  full_name: string;
  private: boolean;
}

interface TokenResponse {
  access_token: string;
  expires_in?: number;
}

export class GitHubError extends Error {}

function encodeBase64Url(value: Uint8Array | string): string {
  const binary = typeof value === "string" ? value : String.fromCharCode(...value);
  return btoa(binary).replaceAll("+", "-").replaceAll("/", "_").replace(/=+$/, "");
}

function decodePem(pem: string): { bytes: Uint8Array; pkcs1: boolean } {
  const pkcs1 = pem.includes("BEGIN RSA PRIVATE KEY");
  const base64 = pem.replace(/-----BEGIN [^-]+-----|-----END [^-]+-----|\s/g, "");
  if (!base64 || (!pkcs1 && !pem.includes("BEGIN PRIVATE KEY"))) {
    throw new Error("GitHub App private key must be an RSA PEM private key");
  }
  return { bytes: Uint8Array.from(atob(base64), (character) => character.charCodeAt(0)), pkcs1 };
}

function derLength(length: number): Uint8Array {
  if (length < 128) {
    return Uint8Array.of(length);
  }
  const bytes: number[] = [];
  for (let remaining = length; remaining > 0; remaining >>= 8) {
    bytes.unshift(remaining & 0xff);
  }
  return Uint8Array.of(0x80 | bytes.length, ...bytes);
}

function der(tag: number, value: Uint8Array): Uint8Array {
  return Uint8Array.of(tag, ...derLength(value.length), ...value);
}

function pkcs1ToPkcs8(pkcs1: Uint8Array): Uint8Array {
  const version = Uint8Array.of(0x02, 0x01, 0x00);
  const rsaAlgorithm = Uint8Array.of(
    0x30,
    0x0d,
    0x06,
    0x09,
    0x2a,
    0x86,
    0x48,
    0x86,
    0xf7,
    0x0d,
    0x01,
    0x01,
    0x01,
    0x05,
    0x00,
  );
  return der(0x30, Uint8Array.of(...version, ...rsaAlgorithm, ...der(0x04, pkcs1)));
}

export async function createAppJwt(
  appId: string,
  privateKey: string,
  now = Math.floor(Date.now() / 1000),
): Promise<string> {
  const decoded = decodePem(privateKey.replaceAll("\\n", "\n"));
  const keyData = new Uint8Array(decoded.pkcs1 ? pkcs1ToPkcs8(decoded.bytes) : decoded.bytes).buffer;
  const key = await crypto.subtle.importKey("pkcs8", keyData, { name: "RSASSA-PKCS1-v1_5", hash: "SHA-256" }, false, [
    "sign",
  ]);
  const header = encodeBase64Url(JSON.stringify({ alg: "RS256", typ: "JWT" }));
  const payload = encodeBase64Url(JSON.stringify({ iat: now - 60, exp: now + 9 * 60, iss: appId }));
  const unsigned = `${header}.${payload}`;
  const signature = await crypto.subtle.sign("RSASSA-PKCS1-v1_5", key, encoder.encode(unsigned));
  return `${unsigned}.${encodeBase64Url(new Uint8Array(signature))}`;
}

async function responseError(response: Response): Promise<GitHubError> {
  let message = response.statusText;
  try {
    const body = (await response.json()) as { message?: string };
    message = body.message || message;
  } catch {
    // GitHub sometimes returns an empty response body.
  }
  return new GitHubError(`GitHub returned ${response.status}: ${message}`);
}

async function githubRequest<T>(path: string, token: string, init: RequestInit = {}): Promise<T> {
  const response = await fetch(`https://api.github.com${path}`, {
    ...init,
    headers: {
      Accept: "application/vnd.github+json",
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
      "User-Agent": "copier-update",
      "X-GitHub-Api-Version": "2022-11-28",
      ...init.headers,
    },
  });
  if (!response.ok) {
    throw await responseError(response);
  }
  return (response.status === 204 ? undefined : await response.json()) as T;
}

export async function exchangeCode(
  clientId: string,
  clientSecret: string,
  code: string,
  verifier: string,
  redirectUri: string,
): Promise<TokenResponse> {
  const response = await fetch("https://github.com/login/oauth/access_token", {
    method: "POST",
    headers: { Accept: "application/json", "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      client_id: clientId,
      client_secret: clientSecret,
      code,
      code_verifier: verifier,
      redirect_uri: redirectUri,
    }),
  });
  const body = (await response.json()) as TokenResponse & { error?: string };
  if (!response.ok || body.error || !body.access_token) {
    throw new GitHubError("GitHub authorization failed");
  }
  return body;
}

export function user(token: string): Promise<GitHubUser> {
  return githubRequest<GitHubUser>("/user", token);
}

async function paginatedCollection<T>(path: string, collection: string, token: string): Promise<T[]> {
  const items: T[] = [];
  for (let page = 1; ; page += 1) {
    const separator = path.includes("?") ? "&" : "?";
    const response = await githubRequest<Record<string, T[]>>(`${path}${separator}page=${page}&per_page=100`, token);
    const pageItems = response[collection];
    items.push(...pageItems);
    if (pageItems.length < 100) {
      return items;
    }
  }
}

export async function installations(token: string): Promise<Installation[]> {
  return paginatedCollection("/user/installations", "installations", token);
}

export async function repositories(token: string, installationId: number): Promise<Repository[]> {
  return paginatedCollection(`/user/installations/${installationId}/repositories`, "repositories", token);
}

export function organizationMembership(
  token: string,
  organization: string,
  username: string,
): Promise<{ role: string; state: string }> {
  return githubRequest(`/orgs/${encodeURIComponent(organization)}/memberships/${encodeURIComponent(username)}`, token);
}

export async function dispatchUpdate(
  appId: string,
  privateKey: string,
  controlRepository: string,
  payload: Record<string, unknown>,
): Promise<void> {
  const repositoryParts = controlRepository.split("/");
  if (repositoryParts.length !== 2 || repositoryParts.some((part) => !part)) {
    throw new Error("CONTROL_REPOSITORY must use owner/repository format");
  }
  const appJwt = await createAppJwt(appId, privateKey);
  const installation = await githubRequest<{ id: number }>(`/repos/${controlRepository}/installation`, appJwt);
  const repositoryName = repositoryParts[1];
  const token = await githubRequest<{ token: string }>(`/app/installations/${installation.id}/access_tokens`, appJwt, {
    method: "POST",
    body: JSON.stringify({ repositories: [repositoryName], permissions: { contents: "write" } }),
  });
  await githubRequest(`/repos/${controlRepository}/dispatches`, token.token, {
    method: "POST",
    body: JSON.stringify({ event_type: "copier-update", client_payload: payload }),
  });
}

File: copier_update_app/github.py
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any
from urllib.parse import quote

import jwt
import requests


class GitHubError(RuntimeError):
    """Raised when a GitHub API request fails."""


@dataclass(frozen=True)
class Repository:
    id: int
    full_name: str
    default_branch: str
    archived: bool = False
    disabled: bool = False


class GitHubAppClient:
    def __init__(
        self,
        app_id: str,
        private_key: str,
        *,
        api_url: str = "https://api.github.com",
        session: requests.Session | None = None,
    ) -> None:
        self.app_id = app_id
        self.private_key = private_key.replace("\\n", "\n")
        self.api_url = api_url.rstrip("/")
        self.session = session or requests.Session()

    def installation_ids(self) -> list[int]:
        installations: list[int] = []
        for item in self._paginate("/app/installations"):
            installations.append(int(item["id"]))
        return installations

    def installation_token(self, installation_id: int, repository_id: int | None = None) -> str:
        body: dict[str, Any] = {}
        if repository_id is not None:
            body["repository_ids"] = [repository_id]
        response = self._request("POST", f"/app/installations/{installation_id}/access_tokens", json=body)
        return str(response["token"])

    def repositories(self, token: str) -> list[Repository]:
        repositories = self._paginate("/installation/repositories", token=token, collection="repositories")
        return [
            Repository(
                id=int(item["id"]),
                full_name=str(item["full_name"]),
                default_branch=str(item.get("default_branch") or "main"),
                archived=bool(item.get("archived", False)),
                disabled=bool(item.get("disabled", False)),
            )
            for item in repositories
        ]

    def copier_answers_file(self, repository: Repository, token: str) -> str | None:
        for filename in (".copier-answers.yaml", ".copier-answers.yml"):
            path = f"/repos/{repository.full_name}/contents/{quote(filename, safe='')}"
            response = self._request("GET", path, token=token, params={"ref": repository.default_branch}, allow_not_found=True)
            if response is not None:
                return filename
        return None

    def has_open_update(self, repository: Repository, token: str, branch_prefix: str) -> bool:
        pulls = self._paginate(
            f"/repos/{repository.full_name}/pulls",
            token=token,
            params={"state": "open", "base": repository.default_branch},
        )
        for pull in pulls:
            head = pull.get("head", {})
            head_repository = head.get("repo") or {}
            if str(head.get("ref", "")).startswith(f"{branch_prefix}-") and head_repository.get("full_name") == repository.full_name:
                return True
        return False

    def create_pull_request(self, repository: Repository, token: str, branch: str, title: str) -> str:
        response = self._request(
            "POST",
            f"/repos/{repository.full_name}/pulls",
            token=token,
            json={
                "base": repository.default_branch,
                "body": "Automated update from the repository's Copier template.",
                "head": branch,
                "title": title,
            },
        )
        return str(response["html_url"])

    def _jwt(self) -> str:
        now = int(time.time())
        return jwt.encode(
            {"iat": now - 60, "exp": now + 9 * 60, "iss": self.app_id},
            self.private_key,
            algorithm="RS256",
        )

    def _paginate(
        self,
        path: str,
        *,
        token: str | None = None,
        collection: str | None = None,
        params: dict[str, Any] | None = None,
    ) -> list[dict[str, Any]]:
        items: list[dict[str, Any]] = []
        page = 1
        while True:
            page_params = dict(params or {})
            page_params.update({"page": page, "per_page": 100})
            response = self._request("GET", path, token=token, params=page_params)
            page_items = response[collection] if collection is not None else response
            items.extend(page_items)
            if len(page_items) < 100:
                return items
            page += 1

    def _request(
        self,
        method: str,
        path: str,
        *,
        token: str | None = None,
        allow_not_found: bool = False,
        **kwargs: Any,
    ) -> Any:
        headers = {
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {token or self._jwt()}",
            "X-GitHub-Api-Version": "2022-11-28",
        }
        response = self.session.request(method, f"{self.api_url}{path}", headers=headers, timeout=30, **kwargs)
        if allow_not_found and response.status_code == 404:
            return None
        if not response.ok:
            try:
                message = response.json().get("message", response.text)
            except ValueError:
                message = response.text
            raise GitHubError(f"{method} {path} returned {response.status_code}: {message}")
        return response.json()
